src/main/scala/com/fulcrumgenomics/bam/IdentifyPrimers.scala

```diff
@@ -0,0 +1,1027 @@
+        .grouped(batchSize)
+        .map { templates => processBatch(templates, metricCounter, readingProgress) }
+    }
+
+    // Write the results
+    val writingProgress = ProgressLogger(this.logger, "written", unit=1000000)
+    outputIterator
+      .flatMap(_.flatMap(_.allReads))
+      .foreach { rec =>
+        writingProgress.record(rec)
+        out += rec
+      }
+
+    val rate = numAlignments.get() / readingProgress.getElapsedSeconds.toDouble
+    logger.info(f"Performed ${numAlignments.get()}%,df full alignments in total ($rate%,.2f alignments/second).")
+    logger.info(f"Wrote ${readingProgress.getCount}%,d records.")
+
+    in.safelyClose()
+    out.close()
+
+
+    // Write metrics
+    // Detailed metrics
+    {
+      val total   = metricCounter.total.toDouble
+      val metrics = metricCounter.map { case (metric, count) =>
+        metric.copy(count = count, frac = count / total)
+      }.toSeq.sortBy(-_.count)
+      Metric.write(PathUtil.pathTo(this.metrics + ".detailed.txt"), metrics)
+    }
+    // Summary metrics
+    Metric.write(PathUtil.pathTo(this.metrics + ".summary.txt"), IdentifyPrimersMetric(metricCounter))
+  }
+
+  /** Creates a new [[InteractiveAlignmentScorer]]. */
+  private def newAligner: InteractiveAlignmentScorer[Primer, SamRecordAlignable] = {
+    InteractiveAlignmentScorer(matchScore, mismatchScore, gapOpen, gapExtend, AlignmentMode.Glocal, this.kswExecutable)
+  }
+
+  // NB: batching alignment inputs to the aligner (i.e. ksw) is empirically faster than given them all or just one at a time.
+  private def completePrimerMatch(matchOption: Option[Either[PrimerMatch, Seq[AlignmentTask[Primer, SamRecordAlignable]]]],
+                                  aligner: InteractiveAlignmentScorer[Primer, SamRecordAlignable],
+                                  alignmentInputBatchSize: Int = 100): Option[PrimerMatch] = {
+    case class PrimerAndScore(primer: Primer, score: Int)
+    import IdentifyPrimers.MaxTwoBy
+    matchOption.flatMap {
+      case Left(primerMatch)      => Some(primerMatch)
+      case Right(alignmentInputs) =>
+        // Add the alignment inputs to the aligner
+        alignmentInputs.grouped(alignmentInputBatchSize).foreach { batch =>
+          aligner.append(batch: _*)
+        }
+        // Get the alignment results
+        val alignmentResults = alignmentInputs.toIterator.zip(aligner.iterator)
+          // Keep results that meet the minimum score and align with a minimum # of query bases
+          .filter { case (alignmentInput, alignmentResult) =>
+            val minQueryEnd = alignmentInput.queryLength - slop
+            alignmentResult.score >= minAlignmentScore && alignmentResult.queryEnd >= minQueryEnd
+          }
+          // now just keep the primer and score
+          .map { case (alignmentInput, alignmentResult) => PrimerAndScore(alignmentInput.query, alignmentResult.score) }
+          // get the best two alignment results by score
+          .maxTwoBy(_.score)
```

Comment:
Why not just `.sort(-_.score).take(2)`?  Your input is already a Seq, so you have it all in memory.